File: Programmers/KaKao/2020_인턴십_보석 쇼핑.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.HashMap;

class Solution {
    public int[] solution(String[] gems) {
        int[] answer = new int[2];
        int gemSize = gems.length;

        HashMap<String, Integer> gemMap = new HashMap<>();
        int indexer = -1;
        for(String gem : gems) {
            if(!gemMap.containsKey(gem)) {
                gemMap.put(gem, ++indexer);
            }
        }

        int gemKinds = gemMap.size();
        int[] gemCount = new int[gemKinds];
        int kindCount = 0;
        int saveSize = Integer.MAX_VALUE;

        Queue<Gem> queue = new LinkedList<>();
        for(int idx = 0; idx < gemSize; idx++) {
            String gem = gems[idx];
            int gemIdx = gemMap.get(gem);

            gemCount[gemIdx]++;
            if(gemCount[gemIdx] == 1) {
                kindCount++;
            }

            queue.add(new Gem(gem, idx));

            if(queue.peek().kind.equals(gem)) {
                Gem temp = queue.peek();
                int tempKind = gemMap.get(temp.kind);
                while(gemCount[tempKind] > 1) {
                    queue.poll();
                    gemCount[tempKind]--;

                    temp = queue.peek();
                    tempKind = gemMap.get(temp.kind);
                }
            }

            if(kindCount == gemKinds) {
                if(idx - queue.peek().index < saveSize) {
                    saveSize = idx - queue.peek().index;
                    answer[0] = queue.peek().index + 1;
                    answer[1] = idx + 1;
                }
            }
        }

        return answer;
    }
    
    class Gem {
        String kind;
        int index;

        Gem(String kind, int index){
            this.kind = kind;
            this.index = index;
        }
    }
}

File: Programmers/KaKao/2020_인턴십_보석 쇼핑.kt
class Solution {
    fun solution(gems: Array<String>): IntArray {
        var answer = intArrayOf()
        var minSize = Int.MAX_VALUE
        val gemsKinds = gems.toSet().size

        var kindCount = 0
        val windowMap:LinkedHashMap<String, Int> = LinkedHashMap()

        for(idx in gems.indices) {
            if(!windowMap.containsKey(gems[idx])) {
                kindCount++
            }
            windowMap.remove(gems[idx])
            windowMap[gems[idx]] = idx

            if(windowMap.size == gemsKinds) {
                val minIdx = windowMap.values.first()
                if(idx - minIdx < minSize) {
                    minSize = idx - minIdx
                    answer = intArrayOf(minIdx + 1, idx + 1)
                }
            }
        }

        return answer
    }
}
